src/main/scala/is/hail/distributedmatrix/DistributedMatrix.scala

```diff
@@ -0,0 +1,114 @@
+package is.hail.distributedmatrix
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.mllib.linalg._
+import org.apache.spark.mllib.linalg.distributed._
+
+import scala.reflect.ClassTag
+
+trait DistributedMatrix[M] {
+  def cache(m: M): M
+
+  def from(rdd: RDD[Array[Double]]): M
+  def from(bm: BlockMatrix): M
+  def from(cm: CoordinateMatrix): M
+
+  def transpose(m: M): M
+  def diagonal(m: M): Array[Double]
+
+  def multiply(l: M, r: M): M
+  def multiply(l: M, r: DenseMatrix): M
+
+  def map4(f: (Double, Double, Double, Double) => Double)(a: M, b: M, c: M, d: M): M
+
+  def map2(f: (Double, Double) => Double)(l: M, r: M): M
+  def pointwiseAdd(l: M, r: M): M
+  def pointwiseSubtract(l: M, r: M): M
+  def pointwiseMultiply(l: M, r: M): M
+  def pointwiseDivide(l: M, r: M): M
+
+  def map(f: Double => Double)(m: M): M
+  def scalarAdd(m: M, i: Double): M
+  def scalarSubtract(m: M, i: Double): M
+  def scalarMultiply(m: M, i: Double): M
+  def scalarDivide(m: M, i: Double): M
+  def scalarAdd(i: Double, m: M): M
+  def scalarSubtract(i: Double, m: M): M
+  def scalarMultiply(i: Double, m: M): M
+  def scalarDivide(i: Double, m: M): M
```

Comment:
Ok division, seems tricky.

Keeping `s / M` and implementing `M / s` as multiplication by multiplicative inverse seems dangerous because of something (admittedly silly) like:
```
scala> Double.MinPositiveValue/Double.MinPositiveValue
res33: Double = 1.0

scala> Double.MinPositiveValue* ( 1/Double.MinPositiveValue)
res34: Double = Infinity
```

Keeping `M / s` seems to suffer the same issue. Namely that multiplicative inverse isn't invertible on small values.